### PR TITLE
docs(options): take ownership of options.txt

### PR DESCRIPTION
**Pull request by lewis6991**
- `src/nvim/options.lua` is now the source of truth
    - generate runtime/lua/vim/_meta/options.lua

Credit to: https://github.com/ii14/emmylua-nvim/tree/master

### CODE REVIEW

**Comment by justinmk** [comment]
Nice! Is the plan here to bootstrap `src/nvim/options.lua` with the docs, as with `eval.lua`? Or alternatively, use `_meta/options.lua` to generate options.txt.

**Comment by lucario387** [comment]
Is there any plan to do the same to `vim.opt` and `vim.opt_local` ? :thinking: 

I think referencing https://github.com/folke/neodev.nvim/blob/main/lua/neodev/build/options.lua for generating it would also be nice

**Comment by lewis6991** [comment]
> Nice! Is the plan here to bootstrap src/nvim/options.lua with the docs, as with eval.lua?

Assuming we are ok to take ownership of options.txt. Since you're asking I'll take that as a yes.

> Is there any plan to do the same to vim.opt and vim.opt_local ? 🤔

Yes.

**Comment by lucario387** [comment]
In the generated file, `vim.o.backupdir` has the value of 
`vim.o.backupdir = ".,/Users/lewrus01/.local/state/nvim/backup//"`

I think `$XDG_*` needs to be gsubbed first

**Comment by lewis6991** [comment]
Reworking this PR to take ownership of the options documentation, similar for what we've done for `builtin.txt`.

**Comment by justinmk** [comment]
> > Is there any plan to do the same to vim.opt and vim.opt_local ? 🤔
> 
> Yes.

Not in favor of that. We shouldn't add more "armor" around vim.opt, we need to deprecate it and enhance vim.o to serve any missing important use-case. https://github.com/neovim/neovim/issues/20107

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -902,9 +902,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 <	Use 'backupdir' to put the backup in a different directory.
 
 						*'backupskip'* *'bsk'*
-'backupskip' 'bsk'	string	(default: "$TMPDIR/*,$TMP/*,$TEMP/*"
-				 Unix: "/tmp/*,$TMPDIR/*,$TMP/*,$TEMP/*"
-				 Mac: "/private/tmp/*,$TMPDIR/*,$TMP/*,$TEMP/*")
+'backupskip' 'bsk'	string	(default "/private/tmp/*,/var/folders/dy/104cjpq539g1y8p4rynvc7tw0000gn/T/*")
```
Comment (lewis6991) [inline runtime/doc/options.txt:905]: todo

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -1583,7 +1578,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 			abbreviations, user commands and the "to" part of the
 			menu commands.  Remove this flag to be able to use a
 			backslash like a CTRL-V.  For example, the command
-			":map X \<Esc>" results in X being mapped to:
+			":map X <Esc>" results in X being mapped to:
```
Comment (lewis6991) [inline runtime/doc/options.txt:1581]: todo

File: runtime/doc/options.txt [hunk 3]
```diff
@@ -1816,9 +1810,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	When Visual mode is active the highlighting isn't used to make it
 	easier to see the selected text.
 
-
-						*'cursorlineopt'* *'culopt'*
-'cursorlineopt' 'culopt' string (default: "number,line")
+					*'cursorlineopt'* *'culopt'*
+'cursorlineopt' 'culopt'string	(default "both")
```
Comment (lewis6991) [inline runtime/doc/options.txt:1814]: todo

File: runtime/doc/options.txt [hunk 4]
```diff
@@ -2718,8 +2708,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	This option cannot be set in a modeline when 'modelineexpr' is off.
 	NOTE: This option is set to "" when 'compatible' is set.
 
-					*'formatlistpat'* *'flp'*
-'formatlistpat' 'flp'	string (default: "^\s*\d\+[\]:.)}\t ]\s*")
+						*'formatlistpat'* *'flp'*
+'formatlistpat' 'flp'	string	(default "^\\s*\\d\\+[\\]:.)}\\t ]\\s*")
```
Comment (lewis6991) [inline runtime/doc/options.txt:2712]: todo

File: runtime/doc/options.txt [hunk 5]
```diff
@@ -2798,8 +2788,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	'errorformat' option: see |errorformat|.
 
 						*'grepprg'* *'gp'*
-'grepprg' 'gp'		string	(default "grep -n ",
-				 Unix: "grep -n $* /dev/null")
+'grepprg' 'gp'		string	(default "grep -n $* /dev/null")
```
Comment (lewis6991) [inline runtime/doc/options.txt:2791]: todo

File: runtime/doc/options.txt [hunk 6]
```diff
@@ -3106,8 +3078,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 <
 
 						*'helpfile'* *'hf'*
-'helpfile' 'hf'		string	(default (MS-Windows) "$VIMRUNTIME\doc\help.txt"
-					 (others) "$VIMRUNTIME/doc/help.txt")
+'helpfile' 'hf'		string	(default "/usr/local/share/nvim/runtime/doc/help.txt")
```
Comment (lewis6991) [inline runtime/doc/options.txt:3081]: todo

File: runtime/doc/options.txt [hunk 7]
```diff
@@ -3432,9 +3401,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	With 'noinfercase' the match is used as-is.
 
 						*'isfname'* *'isf'*
-'isfname' 'isf'		string	(default for Windows:
-			     "@,48-57,/,\,.,-,_,+,,,#,$,%,{,},[,],:,@-@,!,~,="
-			    otherwise: "@,48-57,/,.,-,_,+,,,#,$,%,~,=")
+'isfname' 'isf'		string	(default "@,48-57,/,.,-,_,+,,,#,$,%,~,=")
```
Comment (lewis6991) [inline runtime/doc/options.txt:3404]: todo

File: runtime/doc/options.txt [hunk 8]
```diff
@@ -3483,9 +3450,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	See |option-backslash| about including spaces and backslashes.
 
 						*'isident'* *'isi'*
-'isident' 'isi'		string	(default for Windows:
-					   "@,48-57,_,128-167,224-235"
-				otherwise: "@,48-57,_,192-255")
+'isident' 'isi'		string	(default "@,48-57,_,192-255")
```
Comment (lewis6991) [inline runtime/doc/options.txt:3453]: todo

File: runtime/doc/options.txt [hunk 9]
```diff
@@ -4229,16 +4195,16 @@ A jump table for the options with a short description can be found at |Q_op|.
 	    "g<LeftMouse>"  is "<C-LeftMouse>	(jump to tag under mouse click)
 	    "g<RightMouse>" is "<C-RightMouse>	("CTRL-T")
 
-	      *'mousemoveevent'* *'mousemev'* *'nomousemoveevent'* *'nomousemev'*
-'mousemoveevent' 'mousemev'  boolean	(default off)
+	*'mousemoveevent'* *'mousemev'* *'nomousemoveevent'* *'nomousemev'*
+'mousemoveevent' 'mousemev'boolean	(default off)
```
Comment (lewis6991) [inline runtime/doc/options.txt:4199]: todo

File: runtime/doc/options.txt [hunk 10]
```diff
@@ -4258,9 +4224,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 <	Will make Nvim scroll 5 lines at a time when scrolling vertically, and
 	scroll 2 columns at a time when scrolling horizontally.
 
-					*'mouseshape'* *'mouses'* *E547*
-'mouseshape' 'mouses'	string	(default "i:beam,r:beam,s:updown,sd:cross,
-					m:no,ml:up-arrow,v:rightup-arrow")
+						*'mouseshape'* *'mouses'*
+'mouseshape' 'mouses'	string	(default "")
```
Comment (lewis6991) [inline runtime/doc/options.txt:4228]: todo

File: runtime/doc/options.txt [hunk 11]
```diff
@@ -4785,24 +4748,9 @@ A jump table for the options with a short description can be found at |Q_op|.
...
-					       $XDG_DATA_DIRS[2]/nvim/site/after,
-					       $XDG_DATA_DIRS[1]/nvim/site/after,
-					       $XDG_DATA_HOME/nvim[-data]/site/after,
-					       …
-					       $XDG_CONFIG_DIRS[2]/nvim/after,
-					       $XDG_CONFIG_DIRS[1]/nvim/after,
-					       $XDG_CONFIG_HOME/nvim/after")
+
+					*'runtimepath'* *'rtp'* *vimfiles*
+'runtimepath' 'rtp'	string	(default "/Users/lewis/projects/dotfiles/config/nvim,/etc/xdg/nvim,/Users/lewis/.data/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/usr/local/share/nvim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/lewis/.data/nvim/site/after,/etc/xdg/nvim/after,/Users/lewis/projects/dotfiles/config/nvim/after")
```
Comment (lewis6991) [inline runtime/doc/options.txt:4753]: todo

File: runtime/doc/options.txt [hunk 12]
```diff
@@ -4883,7 +4831,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	height with ":set scroll=0".
 
 						*'scrollback'* *'scbk'*
-'scrollback' 'scbk'	number	(default: 10000)
+'scrollback' 'scbk'	number	(default -1)
```
Comment (lewis6991) [inline runtime/doc/options.txt:4852]: todo?

File: runtime/doc/options.txt [hunk 13]
```diff
@@ -5673,7 +5621,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	The languages are specified with 'spelllang'.
 
 						*'spellcapcheck'* *'spc'*
-'spellcapcheck' 'spc'	string	(default "[.?!]\_[\])'" \t]\+")
+'spellcapcheck' 'spc'	string	(default "[.?!]\\_[\\])'\"\t ]\\+")
```
Comment (lewis6991) [inline runtime/doc/options.txt:5642]: todo

**Comment by lucario387** [comment]
`vim.o.packpath` and `vim.o.runtimepath` currently still have paths that should be gsubbed.

`vim.o.packpath = "/Users/lewis/projects/dotfiles/config/nvim,/etc/xdg/nvim,/Users/lewis/.data/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/usr/local/share/nvim/runtime,/usr/local/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/Users/lewis/.data/nvim/site/after,/etc/xdg/nvim/after,/Users/lewis/projects/dotfiles/config/nvim/after"`

Almost similar thing for `rtp` too

File: .styluaignore [hunk 14]
```diff
@@ -2,3 +2,4 @@
 /src
 /test
 /runtime/lua/vim/re.lua
+/runtime/lua/vim/_meta/options.lua
```
Comment (justinmk) [inline .styluaignore:5]: ?

Comment (lewis6991) [reply .styluaignore:5]: CI is failing on stylua however adding this line doesn't seem to work.

File: runtime/doc/options.txt [hunk 15]
```diff
@@ -1162,15 +1162,15 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 				*'cdhome'* *'cdh'* *'nocdhome'* *'nocdh'*
-'cdhome' 'cdh'		boolean	(default: off)
+'cdhome' 'cdh'		boolean	(default off)
 			global
 	When on, |:cd|, |:tcd| and |:lcd| without an argument changes the
 	current working directory to the |$HOME| directory like in Unix.
 	When off, those commands just print the current directory name.
 	On Unix this option has no effect.
 
-						*'cdpath'* *'cd'* *E344* *E346*
```
Comment (zeertzjq) [inline runtime/doc/options.txt:1172]: missing tags

**Comment by clason** [comment]
language annotations? 🥺

**Comment by lewis6991** [comment]
As with the eval.lua, I'd like to avoid that in this PR just to make the diffs reviewable. The larger the diff, the higher the risk of mistakes not getting caught.

File: runtime/doc/options.txt (hunk 15 above)
Comment (lewis6991) [reply runtime/doc/options.txt:1172]: thanks and fixed

**Comment by lewis6991** [comment]
Unless anyone wants me to hold off, I'll merge this sometime tomorrow.

File: runtime/doc/options.txt [hunk 16]
```diff
@@ -2590,16 +2584,16 @@ A jump table for the options with a short description can be found at |Q_op|.
 	overrule the 'foldlevel' value for specific files.
 	When the value is negative, it is not used.
 
-						*'foldmarker'* *'fmr'* *E536*
```
Comment (zeertzjq) [inline runtime/doc/options.txt:2593]: Missing tag

File: runtime/doc/options.txt [hunk 17]
```diff
@@ -5041,10 +5034,11 @@ A jump table for the options with a short description can be found at |Q_op|.
...
 	If you leave out "options" many things won't work well after restoring
 	the session.
-				*'shada'* *'sd'* *E526* *E527* *E528*
+
+					*'shada'* *'sd'* *E526* *E527* *E528*
 'shada' 'sd'		string	(default for
-				   Win32:  !,'100,<50,s10,h,rA:,rB:
-				   others: !,'100,<50,s10,h)
+                                   Win32:  !,'100,<50,s10,h,rA:,rB:
+                                   others: !,'100,<50,s10,h))
```
Comment (zeertzjq) [inline runtime/doc/options.txt:5041]: Extra parenthesis

File: runtime/doc/options.txt [hunk 18]
```diff
@@ -5264,8 +5258,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	security reasons.
 
 						*'shellquote'* *'shq'*
-'shellquote' 'shq'	string	(default: ""; Windows, when 'shell'
-					contains "sh" somewhere: "\"")
+'shellquote' 'shq'	string	(default ""; Windows, when \'shell\'
```
Comment (zeertzjq) [inline runtime/doc/options.txt:5261]: Extra backslashes

File: src/nvim/options.lua [hunk 19]
```diff
@@ -1,3079 +1,10001 @@
...
+      short_desc = N_('number of command-lines that are remembered'),
+      type = 'number',
+      varname = 'p_hi',
+    },
+    {
+      abbreviation = 'hk',
+      defaults = { if_true = false },
+      full_name = 'hkmap',
+      scope = { 'global' },
+      short_desc = N_('No descripti
```
Comment (zeertzjq) [inline src/nvim/options.lua:7660]:
```suggestion
      defaults = { if_true = '[.?!]\\_[\\])\'" \\t]\\+' },
```

Comment (zeertzjq) [reply src/nvim/options.lua:7660]: Hmm the suggestion is broken again

Comment (lewis6991) [reply src/nvim/options.lua:7660]: What was the suggestion?

File: .luacheckrc [hunk 20]
```diff
@@ -48,4 +48,5 @@ exclude_files = {
   'runtime/lua/vim/_meta/api.lua',
   'runtime/lua/vim/re.lua',
   'src/nvim/eval.lua',
+  'src/nvim/options.lua',
```
Comment (justinmk) [inline .luacheckrc:51]: Do we definitely need to ignore all these files? Generated code can have errors.

File: .styluaignore (hunk 14 above)
Comment (justinmk) [reply .styluaignore:5]: Why is it failing? Is it feasible to fix the warnings?

File: .luacheckrc (hunk 20 above)
Comment (lewis6991) [reply .luacheckrc:51]: Yep, luacheck provides way too many false-negatives in `[[...]]` strings.

Comment (lewis6991) [reply .luacheckrc:51]: Also this file isn't generated.

Comment (lewis6991) [reply .luacheckrc:51]: Maybe, but honestly I've never found luacheck that valuable.

I'll have a look.

**Comment by zeertzjq** [comment]
Needs rebase

File: .luacheckrc (hunk 20 above)
Comment (justinmk) [reply .luacheckrc:51]: hmm. created https://github.com/neovim/neovim/issues/24563

Comment (lewis6991) [reply .luacheckrc]: done

**Comment by bew** [comment]
Will https://github.com/ii14/emmylua-nvim still be necessary with this PR / once nvim 0.10.0 gets released?

**Pull request merged**